### ISSUE TITLE
Inferring spans representing method executions based on a statistical profiler

### ISSUE THREAD

**Issue by felixbarny** [elastic/apm-agent-java#971]
**Motivation**
Our agents mostly capture I/O events like incoming and outgoing HTTP requests. So when most of the time s spend there, it’s easy to troubleshoot latency issues. However, if the application is slow because of inefficient code, users have to manually instrument the application. But in some cases, it’s not feasible to do that, for example, if the code base is huge and it’s unclear which methods cause the slowdown.

The [`trace_methods`](https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-trace-methods) configuration option is often used to match a large portion, if not all methods within a codebase. The documentation warns about the fact that this can significantly increase the overhead and degrade the application's performance.

The slowdown is inherent to the way `trace_methods` works. All methods which match the `trace_methods` expression are instrumented so that a timer is started at the start and stopped at the end of the method. If the execution time was significant enough ([`trace_methods_duration_threshold`](https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-trace-methods)), a span is created. When frequently executing methods are instrumented, it can significantly slow down the application by adding just a little overhead to each invocation. Instrumenting the method can also hinder optimizations the JIT could normally do, like inlining.

**The new approach**
This issue is about adding an alternative to `trace_methods` which does not require to instrument any methods. Instead, a sampling aka statistical profiler would be used as a foundation. These profilers work by gathering the stack trace of the application at frequent intervals, like every 20ms.

By correlating the stack traces with when which span has been active on which thread, we can create a call tree based on the stack traces, correlate them to a span and create spans for it. As the UI is concerned, those spans look just like regular spans so no changes are required in the UI and the APM Server. At a later stage, we could make the UI aware of the profiler-inferred spans and display them with a special icon or color.

**The tradeoffs**
The duration of the spans won’t be as accurate as we’re not exactly measuring the execution time but rather estimate the duration based on the number of consecutive stack traces a method has been present.

To reduce the overhead, the profiler won’t be active all the time. Instead, it’s active for the first 10 seconds of every minute, by default. Only transactions that happen within a profiling session will have profiler-inferred spans.

**Q&A**
* Will `trace_methods` be removed?
  There are no plans to remove that option as it can still be useful in combination or instead of profiler-inferred spans.
* How can I try this out?
  Use a snapshot form https://github.com/elastic/apm-agent-java/pull/972 or build the https://github.com/felixbarny/apm-agent-java/tree/inferred-spans branch.
* How does this relate to https://github.com/elastic/apm/issues/121?
  * The CPU profiling proposal is about having a macro-level view of what the service is doing. This is great to find out about the hotspots of an application. Optimizing those can have a big overall effect on the application. However, it's less useful to troubleshoot latency if the application is mostly idle waiting for I/O.
  * This issue is about creating regular spans for long executing methods. It's mostly used to troubleshoot latency for a specific instance of a transaction.
  * There's no intention to deprecate one in favor of the other. Both are very useful tools to have at your belt to optimize an application.
  * Both can use the same underlying sampling profiler. For the CPU profiling, the stack trace is only considered for threads in a `RUNNABLE` thread state. Also, for CPU profiling, there will be only one flattened data structure, representing a flame graph for the whole profiling session.
* Which settings are there to configure the profiler? See the [docs preview](http://apm-agent-java_983.docs-preview.app.elstc.co/guide/en/apm/agent/java/master/config-profiling.html)

**Screenshot**
<img width="2021" alt="Screen Shot 2019-12-04 at 16 07 52" src="https://user-images.githubusercontent.com/2163464/70901079-5feeec00-1ffa-11ea-9856-b5ffbeeee702.png">

**Comment by felixbarny** [comment]
**How does it work?**
The basis of this is a background thread that gets the stack traces of all the threads we want to profile. The method we’re using for this is `java.lang.management.ThreadMXBean#getThreadInfo(long[] ids, int maxDepth)`.

The other important piece of the puzzle is how these stack traces are correlated with spans. We build on top of `ActivationListener`s which provides a callback for when a span is activated on a specific thread so that it can be retrieved by `ElasticApmTracer.getActive()` to serve as the parent of a span created in a nested method. This is also referred to as in-process context propagation. The `ProfilingActivationListener` creates an `ActivationEvent` each time a sampled span is activated or deactivated on a particular thread. These events contain the nano timestamp (`System.nanoTime()`), the id of the current thread and the `TraceContext`. The events are put into a queue. Each `profiling_sample_rate` (20ms by default), a background thread processes these events and then captures a stack trace of the threads which currently hold an active span. Note that only events that were added since the last sample rate stared are processed. Events added while processing events are left in the queue.

**Event processing logic**
* If activation event
  * If first activation in the current thread
    Create a CallTree.Root and put it in a map with the thread as the key
  * else
    Update the active `TraceContext` in  the `CallTree.Root` for the given thread
    This enables making new `CallTree` nodes a child of the activated span.
* If deactivation event
  * If bottom of the activation stack
    End the `CallTree.Root` and create spans for the relevant `CallTree` elements
  * Else
    Update the active `TraceContext` in  the `CallTree.Root` for the given thread

To reduce the number of spans we create, we only create spans for `CallTree` nodes which have appeared in at least two consecutive samples (`count >= 2`). Also, if a node has only one child and that child has the same `count` we don’t create a span for it but instead add it to the `span.stacktrace` of the leaf node.


**Drawbacks of `getThreadInfo`**
Calling this API causes the JVM to enter a global safe point. That means that all threads have to wait for all other threads to enter a safe point, which causes a stop the world pause. See also http://psy-lob-saw.blogspot.com/2016/02/why-most-sampling-java-profilers-are.html and https://stackoverflow.com/questions/59171277/does-requesting-a-stack-trace-from-a-particular-thread-require-all-threads-to-be.

I did some measurements with our current benchmarks which measure the time it takes to record one HTTP transaction and one JDBC span. Executing with 4 concurrent threads on my  MacBook Pro (15-inch, 2017, 2.9GHz), the latency per transaction went up ~0-1.5µs on average and from ~40-~400µs for the 99.9th percentile. It’s very hard to draw generic conclusions from it because it depends on the application how long it takes for all threads to reach a safe point. 

**Future work**
I also did some investigation on how we could integrate [async-profiler](https://github.com/jvm-profiling-tools/async-profiler) which does not suffer from safepoint-bias and also allows for wall clock profiling. There are some obstacles but in general, it seems feasible and worth it.


**Caveats**
* A regular span can’t be a child of a profiler-inferred span
* A profiler-inferred span might be classified as a child of a regular span when it actually is a parent of that span

<img width="477" alt="Screen Shot 2019-12-04 at 17 23 54" src="https://user-images.githubusercontent.com/2163464/70909241-50c56980-200d-11ea-82fb-66f0a4ae4765.png">

This happens when both start at almost the same time. In this example, `DispatcherServlet#render` directly calls `View#render`. The former one is captured by the profiler, the latter one by instrumenting the method. But only after the `View#render` span is created and activated, we see the stack trace containing `DispatcherServlet#render` the first time and thus assume it has been started afterwards, therefore further up in the stack. But `DispatcherServlet#render` actually takes longer than `View#render` as in reality it’s further down the stack and does some processing after `View#render` is done.


**Open questions/TODOs**
- Filter by application_packages by default?
    - Disable if not set?
    - Users might be interested in non-application methods
- Handling of activation events
    - Buffer activation events in memory-mapped file?
        - Mainly necessary if we want to support async-profiler as we’ll only process events after the profiling session has ended.
        - Can also be useful for the current profiler in order to not lose events.
    - Ring Buffer for activation events/object pooling
    - Can an activation event with a later timestamp be in the queue before an earlier one?
        - Probably yes…
        - Currently, we rely on that not being the case for `processActivationEventsUpTo(long nanoTimestamp)`
- Align configuration options with @axw’s profiler settings
    - `ELASTIC_APM_CPU_PROFILE_INTERVAL`
    - `ELASTIC_APM_CPU_PROFILE_DURATION`

**Comment by eyalkoren** [comment]
Great writeup! 
[See suggestion](https://github.com/elastic/apm-agent-java/pull/972/files#r358683449) that maybe can fix the parent-child problem.

**Comment by Yogi-SAP** [comment]
Thanks for this. A few additional questions/points apart from the one discussed yesterday,

1. How will RUM's sampling frequency impact the configuration of profiling window? Will it override? If the profiling window of the stat profiler is 1 second and the RUM is 100% which one overrides which?
2. Could you please attach a snapshot JAR?

**Comment by Yogi-SAP** [comment]
Please publish the overhead when you have 100 threads. Info on both the safe point delay and the background thread overhead will be helpful.

**Comment by felixbarny** [comment]
Hi Yogi,

the `transaction_sample_rate` is not related in any way with `profiling_sampling_duration`. The former is about the percentage of sampled transactions, the latter the delay between two thread dumps.

I've uploaded a snapshot here: https://github.com/elastic/apm-agent-java/pull/972#issuecomment-567018675

**Comment by felixbarny** [comment]
> Please publish the overhead when you have 100 threads.

While we plan to do load tests before releasing the feature, it will take us some time to set up a proper environment etc.

Feel free to do some load testing in your own environment. But be aware that this is not an official build, just a snapshot of the current development.

**Comment by WassimDhib** [comment]
Hi, 

This is a very good news.
I was already using the trace_methods option, and it's true that it can cause significant overhead when diving into low level frameworks.

I see youre using micro benchmarking (JMH), I'm interested in doing macro benchmarks to measure overhead. 
Would you like to see the results ?

**Comment by felixbarny** [comment]
Update: I was able to integrate async-profiler with the agent 🎉 

Async-profiler has much lower overhead than `ThreadMXBean#getThreadInfo` and does not rely on safepoints.

The biggest downside is that async-profiler does not work on Windows. I think it still makes sense to only support async-profiler and not have a fallback to `ThreadMXBean#getThreadInfo` on Windows. The time to reach a safe point (which means a stop-the-world pause for the application) is quite unpredictable and can regularly be as high as 5ms.

How it works in a nutshell:
- The agent starts async-profiler in wallclock mode for 10s and lets it write to a JFR (Java Flight Recorder format) file.
- During that profiling session, the agent captures span activation events and writes them to a ring buffer in a garbage-free way.  
- The events in the ring buffer are consumed by a thread that writes them into a memory-mapped file (MappedByteBuffer) which has a size of around 10mb which can hold around 100k activation events.
- After the profiling session is over, the binary JFR file is consumed using a MappedByteBuffer (doesn't consume heap memory proportional to the file size) and activation events are correlated with stack traces. This is mostly garbage free as well. The only source of allocations is we have to sort the execution sample events, consisting of a timestamp, threadId, and stackTraceId. We have to do this so we can correlate the already sorted activation events with the stack traces.

TODOs
- [ ]  Correlation with Java threads is not working seamlessly at the moment
  There's a new, unreleased feature for that but there are some problems https://github.com/jvm-profiling-tools/async-profiler/issues/277#issuecomment-569662985

Gotchas
- The actual sampling interval is lower than what's provided to async-profiler. For example, when configuring async-profiler to take a thread snapshot of all threads every 5ms, the actual sampling rate is more like 25ms.

--- 

> I see you are using micro benchmarking (JMH), I'm interested in doing macro benchmarks to measure overhead.
Would you like to see the results ?

Sure!

**Comment by WassimDhib** [comment]
Hi

can you provide a new SNAPSHOT jar with async-profiler ?

I tested the previous one (1.12.1), I can see app.inferred spans but I have no full gc events, is this normal ?

**Comment by Yogi-SAP** [comment]
hi Felix, what would be shipping out with the 7.7 version? The asynch based one or the one with the earlier ThreadJMX bean one? I am not sure I completely understand which one is going to be the one we have to start testing and expect for our 7.7 version release.

**Comment by felixbarny** [comment]
@Yogi-SAP 
The APM agents are not tied to the stack release schedule so this feature can be released independently. However, we don't have a specific target date for it yet. There are still some uncertainties and we might need UI and server changes. Those are tied to the stack release schedule but maybe we can make it work even with older stack versions. The plan is to only ship the async-profiler branch and to abandon the other one. Would that work for you or do you have services running on Windows?

If you want to get information about release dates, please get in touch with our PMs.

@WassimDhib 
What is it you are testing there?
> I tested the previous one (1.12.1), I can see app.inferred spans but I have no full gc events, is this normal ?

Without knowing what your benchmark does it's hard to tell 🙂

**Comment by felixbarny** [comment]
Here's a snapshot build with async-profiler: https://github.com/elastic/apm-agent-java/pull/983#issuecomment-573329916

**Comment by felixbarny** [comment]
I did some load tests with 100 concurrent threads once with continuous profiling and once without (but agent still enabled).

The configuration was:

```properties
profiling_spans_enabled=true
profiling_interval=10s
profiling_duration=10s
profiling_sampling_interval=5ms
```

I could not measure any significant impact on throughput.

At a later stage, we are planning more extensive load tests that run nightly on dedicated hardware. We can then get more detailed information about the impact on allocation rate, GC count and time, CPU, latency, startup time, and heap utilization.

**Comment by felixbarny** [comment]
Another snapshot, including the `profiling_spans_min_duration` configuration option and async-profiler 1.7-ea: [elastic-apm-agent-1.12.1-SNAPSHOT.jar.zip](https://github.com/elastic/apm-agent-java/files/4130293/elastic-apm-agent-1.12.1-SNAPSHOT.jar.zip)

**Comment by Yogi-SAP** [comment]
> I did some load tests with 100 concurrent threads once with continuous profiling and once without (but agent still enabled).
> 
> The configuration was:
> 
> ```ini
> profiling_spans_enabled=true
> profiling_interval=10s
> profiling_duration=10s
> profiling_sampling_interval=5ms
> ```
> 
> I could not measure any significant impact on throughput.
> 
> At a later stage, we are planning more extensive load tests that run nightly on dedicated hardware. We can then get more detailed information about the impact on allocation rate, GC count and time, CPU, latency, startup time, and heap utilization.

hi Felix, thanks for this but if you notice the timing of the ask, it was for when the JMX threadJMX based profiler. With the new asynch profiler it no longer holds relevance. As you noted, there will be detailed performance numbers from your side, we are looking at guidelines on what the expected overhead is across parameters like memory, CPU and others. Like how some of the APM solutions provide guidelines on how they performance what overhead is expected.

**Comment by felixbarny** [comment]
Please hold off with performance tests for now. I'm doing some optimizations and will upload a new snapshot that is ready to test soon.

**Comment by felixbarny** [comment]
A new snapshot, including async-profiler 1.7.0-ea3:  [elastic-apm-agent-1.13.1-SNAPSHOT.jar.zip](https://github.com/elastic/apm-agent-java/files/4220156/elastic-apm-agent-1.13.1-SNAPSHOT.jar.zip)


Some highlights:
- Much reduced memory allocations
- Faster processing of large JFR files
- Faster processing in general
- Only profiles threads that are actively running a transaction
- The `profiling_interval` is now stable, irrespective of the number of threads
- Decrease time-to-safepoint by switching from MappedByteBuffer to DirectByteBuffer + FileChannel
- Bug fix that can prematurely end call tree nodes

happy performance testing 🙂

**Comment by Yogi-SAP** [comment]
hi Felix, could you please check the configuration document link? i am getting a 404 status when accessing the page.

**Comment by felixbarny** [comment]
I have updated the link in the issue description.

**Comment by wolframhaussig** [comment]
Hi Felix, I just wanted to try it out and installed it on our test system. Now I have a question that might be a bit dumb but: How do I know if the profiling is working? I have enabled TRACE logging but do not see any information about the profiling...

**Comment by felixbarny** [comment]
To enable it, set `profiling_spans_enabled` to `true`. By default, it will profile for 10s and then sleep for 51s. If you want continuous profiling, set `profiling_interval=10s` and `profiling_duration=10s`. I'm still a bit undecided whether the continuous profiling might be the better default.

On debug log level, you should then see logs like this:
```
Start profiling session
Processing traces took {}µs
End profiling session
```

This will create spans with `span.subtype: inferred`.

**Comment by wolframhaussig** [comment]
I've got it to work - after restarting the application. Is it possible that either the reload of the setting does not work correctly or that the documentation incorrectly states that profiling_spans_enabled is dynamic? I have started the application with your snapshot and afterwards modified the elasticapm.properties in the same directory to enable profiling.

But after I got it to work I think: 👍

**Comment by felixbarny** [comment]
The `profiling_spans_enabled` option is in fact dynamic. My guess is that you're loading `elasticapm.properties` from the classpath, in which case it can't be reloaded. To work around this in a dev environment, set `-Delastic.apm.config_file=/path/to/elasticapm.properties`
